cheroot/server.py

```diff
@@ -698,42 +705,73 @@ def read_request_line(self):
+            path = (uri
+                    # https://tools.ietf.org/html/rfc7230#section-5.3.4
+                    if self.proxy_mode or uri == ASTERISK
+                    else urllib.parse.urlsplit(uri).path)
+        elif self.method == 'CONNECT':
+            if not self.proxy_mode:
+                self.simple_response('400 Bad Request')
+                return False
 
-        # Unquote the path+params (e.g. "/this%20path" -> "/this path").
-        # http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2
-        #
-        # But note that "...a URI must be separated into its components
-        # before the escaped characters within those components can be
-        # safely decoded." http://www.ietf.org/rfc/rfc2396.txt, sec 2.4.2
-        # Therefore, "/this%2Fpath" becomes "/this%2Fpath", not "/this/path".
-        try:
-            atoms = [
-                unquote_to_bytes(x)
-                for x in quoted_slash.split(path)
-            ]
-        except ValueError as ex:
-            self.simple_response('400 Bad Request', ex.args[0])
-            return False
-        path = b'%2F'.join(atoms)
+            # https://tools.ietf.org/html/rfc7230#section-5.3.3
+            authority = uri
+        else:
+            # https://tools.ietf.org/html/rfc7230#section-5.3.1 (origin_form) and
+            # https://tools.ietf.org/html/rfc7230#section-5.3.2 (absolute form)
+            scheme, authority, path, qs, fragment = urllib.parse.urlsplit(uri)
+
+            if (self.strict_mode and not self.proxy_mode) and (scheme or authority):
+                self.simple_response('400 Bad Request',
+                                     'Absolute URI not allowed if server is not a proxy.')
+                return False
+
+            if fragment:
+                self.simple_response('400 Bad Request',
+                                     'Illegal #fragment in Request-URI.')
+                return False
+
+            if path is None:
+                self.simple_response('400 Bad Request',
+                                     'Invalid path in Request-URI.')
+                return False
+
+            # Unquote the path+params (e.g. "/this%20path" -> "/this path").
+            # http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2
+            #
+            # But note that "...a URI must be separated into its components
+            # before the escaped characters within those components can be
+            # safely decoded." http://www.ietf.org/rfc/rfc2396.txt, sec 2.4.2
+            # Therefore, "/this%2Fpath" becomes "/this%2Fpath", not "/this/path".
+            try:
+                atoms = [
+                    unquote_to_bytes(x)
+                    for x in quoted_slash.split(path)
+                ]
+            except ValueError as ex:
+                self.simple_response('400 Bad Request', ex.args[0])
+                return False
+            path = b'%2F'.join(atoms)
+
+        if path[0] != b'/':
```

Comment:
Wow! It turned out that `path[0]` returns `int`, not `bytes`, which is `47` for `/`, thus prepending `/` unconditionally.

(just a note)